spyderlib/restart_app.py

```diff
@@ -129,18 +180,60 @@ def main():
     shell = os.name != 'nt'
 
     # Wait for original process to end before launching the new instance
-    while True:
+    max_counter = 1500  # Max loops to perform before aborting
+    counter = 0
+    while counter < max_counter:  # This is equivalent to ~ 5 minute (1500*0.2)
         if not is_pid_running(pid):
             break
         time.sleep(0.2)  # Throttling control
+        counter += 1
+
+    if counter == max_counter:
+        error_message(type_='close')
 
     env = os.environ.copy()
-    try:
-        subprocess.Popen(command, shell=shell, env=env)
-    except Exception as error:
-        print(command)
-        print(error)
-        time.sleep(15)
+
+    reset_ok = True
+
+    # Reset spyder if needed
+    # -------------------------------------------------------------------------
+    if reset:
+        # Build reset command
+        command_reset = '"{0}" "{1}" {2}'.format(python, spyder, args_reset)
+
+        try:
+            p = subprocess.Popen(command_reset, shell=shell)
+            p.communicate()
+            pid_reset = p.pid
+        except Exception as error:
+            p.kill()
+            error_message(type_='reset', error=error)
+
+        counter = 0
+        max_counter = 50
+        # Wait for reset process to end before restarting
+        while counter < max_counter:  # This is ~= 10 seconds (50*0.2)
+            if not is_pid_running(pid_reset):
+                break
+            time.sleep(0.2)  # Throttling control
+            counter += 1
+
+        if counter == max_counter:
+            # The rest subprocess took too long and it is killed
+            p.kill()
+            reset_ok = False
+
+    # Restart
+    # -------------------------------------------------------------------------
+    if reset_ok:
+        # Try to restart
+        try:
+            p = subprocess.Popen(command, shell=shell, env=env)
+        except Exception as error:
+            p.kill()
+            error_message(type_='restart', error=error)
+    else:
+        error_message(type_='reset')

```

Comment:
Cause I have two places to define `reset_ok = False`, one for an exception and the other for timeout... so I guess this looked cleaner to me as opposed, to writing `error_message(type_='reset')` in both places (where I define `reset_ok = False`)